indexing-solr/src/main/java/edu/unc/lib/boxc/indexing/solr/filter/SetDatastreamFilter.java

```diff
@@ -227,41 +229,56 @@ private void addDerivatives(List<Datastream> dsList, PID pid, boolean ownedByOth
+            String name = type.getId();
+            String mimetype = type.getMimetype();
+            String extension = type.getExtension();
+            File file = deriv.getFile();
+            Long filesize = file.length();
+            String filename = file.getName();
+            var derivative = new DatastreamImpl(owner, name, filesize, mimetype, filename, extension, null, null);
+            addDerivativeToList(dsList, derivative);
         });
     }
 
     /**
      * Used to selectively add only thumbnail datastreams
      *
-     * @param contentObject should be the work object with the thumbnail relation
+     * @param workObject the work object with the thumbnail relation
      * @param datastreams work object's datastreams to add thumbnail streams to
      */
-    private void addThumbnailDerivatives(ContentObject contentObject, List<Datastream> datastreams) {
-        WorkObject workObj = (WorkObject) contentObject;
-        FileObject thumbnailObject = workObj.getThumbnailObject();
+    private void addThumbnailDerivatives(WorkObject workObject, List<Datastream> datastreams) {
+        FileObject thumbnailObject = workObject.getThumbnailObject();
 
         if (thumbnailObject != null) {
-            List<DatastreamType> types = new ArrayList<>();
-            types.add(DatastreamType.THUMBNAIL_SMALL);
-            types.add(DatastreamType.THUMBNAIL_LARGE);
-
-            addDerivatives(datastreams, thumbnailObject.getPid(), true, types);
+            var updatedDatastreams = clearPreviousThumbnailDatastreams(datastreams);
+            addDerivatives(updatedDatastreams, thumbnailObject.getPid(), true, THUMBNAIL_DS_TYPES);
         }
     }
 
-    private void addToDatastreamList(DatastreamType type, String owner, List<Datastream> dsList, File file) {
-        String name = type.getId();
-        String mimetype = type.getMimetype();
-        String extension = type.getExtension();
-
-        Long filesize = file.length();
-        String filename = file.getName();
+    /**
+     *  There may be thumbnail streams from the primary object, so we'll clear those
+     *  before adding the assigned thumbnail datastreams
+     *
+     * @param datastreams full list of datastreams to index for the work object
+     * @return modified list of datastreams without thumbnail datastreams
+     */
+    private List<Datastream> clearPreviousThumbnailDatastreams(List<Datastream> datastreams) {
+        for (Datastream datastream : datastreams) {
+            var type = DatastreamType.getByIdentifier(datastream.getName());
+            if (THUMBNAIL_DS_TYPES.contains(type)) {
+                datastreams.remove(datastream);
+            }
+        }
+        return datastreams;
+    }
 
-        dsList.add(new DatastreamImpl(owner, name, filesize, mimetype, filename, extension, null, null));
+    private void addDerivativeToList(List<Datastream> dsList, DatastreamImpl derivative) {
+        dsList.add(derivative);
```

Comment:
I think it'd make sense to remove this helper method at this point since it is just wrapping a `.add` call

Reply:
I'll rework the methods to streamline it a bit more